v2/blockstore/readwrite.go

```diff
@@ -57,6 +57,9 @@ func WithIndexPadding(p uint64) Option {
 	}
 }
 
+// TODO When we are constructing a UnxFS DAG from a user file using this read-write blockstore as the staging store,
+// we might not have the root Cid will we finalise.
+// Please can we add an option to skip the root here and give it when we finalize the DAG ?
```

Comment:
That requires a full re-location / re-write of the car on finalize though.

Do you know there's a single root CID and it's expected hash type? we'll probably need to do something tricky to avoid the change in length of the header that'll come from filling in the true root cid at the end.

Reply:
It really won't matter: when you construct a dag you tart "leaves first", yet the car file for Fil-purposes must be "leaves last". Since you will be literally rewriting the position of every single block in your .car, you might as well start from he header: that is you don't *need* to know how large it is.

Reply:
@willscott @mvdan This is not a must-have. We can work our way around it by:

- Generate the UnixFS DAG using a Badger blockstore to get the root Cid.
- Generate the UnixFS DAG yet again using a CARv2 read-write blockstore now that we have the root.
- Delete the Badger blockstore.

However , this has obvious inefficiencies and we wont be able to remove _completely_ from deal making.

Reply:
@ribasushi seems to be saying that data transfer is not in the canonical selector order. that would mean needing a full re-write on finalization. Can you confirm, @aarshkshah1992  if that will be true? 

If it's just that you don't know the root at the beginning, but will at the end, it's probably better for your code to have the hack to make this work: you'll start by specifying a single root of a 0 cid of the expected multi-hash type, and then we'll let you update/replace it at the time of finalization to the real root.

Reply:
Wait... no. Grapsync *should* be doing stuff in "tarversal" order ( there is some internal buffering stuff, but that should be hidden from you... I hope )
The context was "When we are constructing a UnxFS DAG from a user file" which to me implies the local `lotus client import`. Did I misunderstand?

Reply:
Ok. great. this issue of roots is in the context (i think) of the receiving side of a graphsync transfer.
I don't think we're using any of this on the client where the import happens.

Reply:
(at least for this first draft)

Reply:
> this issue of roots is in the context (i think) of the receiving side of a graphsync transfer.

Graphsync always knows the root somehow ( it needs to start validation-traversal from somewhere ). You definitely should have the root cid available by some means before you received your first ever block ( to validate that this first block is what you expect, and so on )

Reply:
We'll talk about this in the daily carv2 sync :)

Reply:
We can deprioritize this as per today's call.

Reply:
Had a chat with @raulk today and re-opening this as removing Badger completely from all aspects of Deal making is an important goal of the CARv2 blockstore project which we will not be able to accomplish without this feature.

Reply:
@aarshkshah1992 please open a new issue instead, with the notes you mentioned during our daily sync. Tracking stuff to do via draft PR sub-threads is not going to work :)